pyro/ops/integrator.py

```diff
@@ -83,7 +83,7 @@ def potential_grad(potential_fn, z):
         else:
             raise e
 
-    grads = grad(potential_energy, z_nodes)
+    grads = grad(potential_energy, z_nodes, retain_graph=True)
```

Comment:
@fritzo, @fehiepsi - Ref to our discussion in #2283. I had to do this by default in the integrator, since this issue is not specific to HMC initialization, and we bump into the same issue when a proposal might be rejected at any point. Unless `retain_graph` is `True` we will end up hitting the cache but the cached value's `grad` would have been cleared. Storing the graph for all samples is not needed, but it is hard to say apriori when we need to do this. I didn't see any significant memory usage increase at least on our tests and examples, but let me know if this seems like a dicey change.

Reply:
I just worry that this will retain graphs of every leap_frog step. From #2283, if this just happens at `find_reasonable_step_size`, then it might be better to clone the variable in this function.

Reply:
I think we can get away by just cloning in one part of the codebase, I'll make that change. We are not making any in place updates in the leapfrog integrator, but I agree that its best to avoid `retain_graph=True`.

Reply:
I was able to resolve this by cloning. I still face an issue when trying to jit compile where the tracer complains that it cannot add a constant tensor with `requires_grad=True`; there's an xfailing test for that.